Projects/Server/Gumps/New/DynamicGridGump.cs

```diff
@@ -0,0 +1,128 @@
+        protected const int ArrowLeftHeight = 16;
+        protected const int ArrowRightID1 = 0x15E1;
+        protected const int ArrowRightID2 = 0x15E5;
+        protected const int ArrowRightWidth = 16;
+        protected const int ArrowRightHeight = 16;
+
+        private readonly int _x;
+        private readonly int _y;
+        private readonly GumpFlags _flags;
+
+        protected virtual ushort BorderSize => 10;
+        protected virtual ushort OffsetSize => 1;
+        protected virtual ushort EntryHeight => 20;
+        protected virtual ushort OffsetGumpId => 0x0A40;
+        protected virtual ushort HeaderGumpId => 0x0E14;
+        protected virtual ushort EntryGumpId => 0x0BBC;
+        protected virtual ushort BackGumpId => 0x13BE;
+        protected virtual ushort TextHue => 0;
+        protected virtual ushort TextOffsetX => 2;
+
+        protected DynamicGridGump(int x, int y, GumpFlags flags = GumpFlags.None)
+        {
+            _x = x;
+            _y = y;
+            _flags = flags;
+        }
+
+        protected static int GetButtonID(int typeCount, int type, int index)
+        {
+            return 1 + index * typeCount + type;
+        }
+
+        protected static bool SplitButtonID(int buttonID, int typeCount, out int type, out int index)
+        {
+            if (buttonID < 1)
+            {
+                type = 0;
+                index = 0;
+                return false;
+            }
+
+            buttonID -= 1;
+            index = Math.DivRem(buttonID, typeCount, out type);
+
+            return true;
+        }
+
+        public override void SendTo(NetState ns)
+        {
+            GridGumpBuilder<StaticStringsHandler> builder = new(_flags, BorderSize, OffsetSize, EntryHeight,
+                OffsetGumpId, HeaderGumpId, EntryGumpId, BackGumpId, TextHue, TextOffsetX);
+
+            try
+            {
+                Build(ref builder);
+                ns.AddGump(this);
+                Send(ns, in builder);
+            }
+            finally
+            {
+                builder.Dispose();
+            }
+        }
+
+        protected abstract void Build(ref GridGumpBuilder<StaticStringsHandler> builder);
+
+        private void Send(NetState ns, in GridGumpBuilder<StaticStringsHandler> builder)
+        {
+            ref readonly StaticStringsHandler stringsWriter = ref builder.StringsWriter;
+
+            int worstLayoutLength = Zlib.MaxPackSize(builder.LayoutSize);
+            int worstStringsLength = Zlib.MaxPackSize(stringsWriter.BytesWritten);
+
+            int maxLength = 40 + worstLayoutLength + worstStringsLength;
+
+            SpanWriter writer = new(maxLength);
+            writer.Write((byte)0xDD); // Packet ID
+            writer.Seek(2, SeekOrigin.Current);
+
+            writer.Write(Serial);
+            writer.Write(TypeID);
+            writer.Write(_x);
+            writer.Write(_y);
+
+            builder.FinalizeLayout();
```

Comment:
All of the calls are throwing a warning saying the method is impure and executed against a readonly ref, causing a automatic defensive copy.

Reply:
Good catch: I fixed all the wrong readonly ref  parameters for new gumps